File: domain/src/main/java/com/parabola/domain/interactor/SearchInteractor.java
package com.parabola.domain.interactor;

import static java.util.Collections.emptyList;
import static java.util.Collections.unmodifiableList;

import com.parabola.domain.model.Album;
import com.parabola.domain.model.Artist;
import com.parabola.domain.model.Playlist;
import com.parabola.domain.model.Track;

import java.util.List;

import io.reactivex.Single;

public abstract class SearchInteractor {

    protected static final int ARTISTS_SEARCH_MAX_LIMIT = 3;
    protected static final int ALBUMS_SEARCH_MAX_LIMIT = 5;
    protected static final int TRACKS_SEARCH_MAX_LIMIT = 20;
    protected static final int PLAYLIST_SEARCH_MAX_LIMIT = 5;


    public final Single<SearchResult> search(String query) {
        return Single.zip(searchArtists(query), searchAlbums(query), searchTracks(query), searchPlaylists(query), SearchResult::new)
                .onErrorReturnItem(SearchResult.EMPTY_SEARCH_RESULT);
    }


    public abstract Single<List<Artist>> searchArtists(String query);

    public abstract Single<List<Album>> searchAlbums(String query);

    public abstract Single<List<Track>> searchTracks(String query);

    public abstract Single<List<Playlist>> searchPlaylists(String query);


    public static final class SearchResult {
        private static final SearchResult EMPTY_SEARCH_RESULT
                = new SearchResult(emptyList(), emptyList(), emptyList(), emptyList());

        public final List<Artist> artists;
        public final List<Album> albums;
        public final List<Track> tracks;
        public final List<Playlist> playlists;

        private SearchResult(List<Artist> artists, List<Album> albums, List<Track> tracks, List<Playlist> playlists) {
            this.artists = unmodifiableList(artists);
            this.albums = unmodifiableList(albums);
            this.tracks = unmodifiableList(tracks);
            this.playlists = unmodifiableList(playlists);
        }
    }

}

File: search_feature/src/main/java/com/parabola/search_feature/SearchInteractor.kt
package com.parabola.search_feature

import com.parabola.domain.model.Album
import com.parabola.domain.model.Artist
import com.parabola.domain.model.Playlist
import com.parabola.domain.model.Track
import io.reactivex.Single

val EMPTY_SEARCH_RESULT = SearchResult(emptyList(), emptyList(), emptyList(), emptyList())

const val ARTISTS_SEARCH_MAX_LIMIT = 3L
const val ALBUMS_SEARCH_MAX_LIMIT = 5L
const val TRACKS_SEARCH_MAX_LIMIT = 20L
const val PLAYLIST_SEARCH_MAX_LIMIT = 5L

abstract class SearchInteractor {
    fun search(query: String): Single<SearchResult> {
        return Single.zip(
            searchArtists(query),
            searchAlbums(query),
            searchTracks(query),
            searchPlaylists(query)
        ) { artists: List<Artist>, albums: List<Album>, tracks: List<Track>, playlists: List<Playlist> ->
            SearchResult(
                artists,
                albums,
                tracks,
                playlists
            )
        }
            .onErrorReturnItem(EMPTY_SEARCH_RESULT)
    }

    abstract fun searchArtists(query: String): Single<List<Artist>>
    abstract fun searchAlbums(query: String): Single<List<Album>>
    abstract fun searchTracks(query: String): Single<List<Track>>
    abstract fun searchPlaylists(query: String): Single<List<Playlist>>

}
